lib/resultuploaders/s3compatible/s3compatible.rb

```diff
@@ -0,0 +1,199 @@
+          # such as Alibaba Cloud OSS, which do not support virtual-hosted style.
+          # Forces path-style access to ensure proper functionality.
+        )
+        @bucket = @s3_resource.bucket(@bucket_name)
+        @logger.info("S3Compatible bucket connected: #{@bucket_name}")
+        true
+      end
+    end
+
+    def create_project_folder
+      handle_exceptions(__method__) do
+        @path = "#{@repo}/CI/#{@tag}-#{@timestamp}"
+        update_index_file
+        @url = generate_url("#{@path}/#{INDEX_FILE_NAME}")
+        @logger.info("S3Compatible project folder created: #{@url}")
+        true
+      end
+    end
+
+    def upload_file(l_path, r_name)
+      handle_exceptions(__method__) do
+        remote_path = "#{@path}/#{r_name}"
+        type = 'text/html' if r_name.end_with?('.html')
+        obj = @bucket.object(remote_path)
+        obj.upload_file(l_path, content_type: type)
+        update_index_file
+        @logger.info("S3Compatible file uploaded: #{remote_path}")
+        true
+      end
+    end
+
+    def update_file_content(content, r_name)
+      handle_exceptions(__method__) do
+        remote_path = "#{@path}/#{r_name}"
+        obj = @bucket.object(remote_path)
+        obj.put(body: content)
+        @logger.info("S3Compatible file content updated: #{remote_path}")
+        true
+      end
+    end
+
+    def delete_file(r_name)
+      handle_exceptions(__method__) do
+        r_path = "#{@path}/#{r_name}"
+        obj = @bucket.object(r_path)
+        obj.delete
+        update_index_file
+        @logger.info("S3Compatible file deleted: #{r_path}")
+        true
+      end
+    end
+
+    def close; end
+
+    private
+
+    def update_index_file
+      title = @path.split('/').last
+      html_content = build_html_content(title)
+      html_content += list_objects
+      save_index_file(html_content)
+    end
+
+    def build_html_content(title)
+      <<~HTML
+        <!DOCTYPE html>
+        <html lang="en">
+        <head>
+            <meta charset="UTF-8">
+            <meta name="viewport" content="width=device-width, initial-scale=1.0">
+            <title>#{title}</title>
+            <style>
+                body {
+                    font-family: Arial, sans-serif;
+                }
+                h1 {
+                    color: #333;
+                }
+                ul {
+                    list-style-type: none;
+                    padding: 0;
+                }
+                li {
+                    margin: 5px 0;
+                }
+                a {
+                    text-decoration: none;
+                    color: #007bff;
+                }
+                a:hover {
+                    text-decoration: underline;
+                }
+            </style>
+        </head>
+        <body>
+            <h1>#{title}</h1>
+            <ul>
+      HTML
+    end
+
+    def list_objects
+      objects_html = ''
+      objects = @bucket.objects(prefix: @path)
```

Comment:
Please do not use `@bucket.objects(prefix: @path)`, which incurs extra traffic. Remember objects when uploading instead.